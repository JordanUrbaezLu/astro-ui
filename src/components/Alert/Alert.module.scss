@use "../../utility/globals";

.alertContainer {
  align-items: center;
  border-radius: 8px;
  column-gap: 10px;
  display: flex;
  justify-content: space-between;
  min-width: 250px;
  padding: 10px 12px;
  width: fit-content;

  &.error {
    background-color: rgb(255, 210, 210);
    border: 3px solid red;
  }

  &.info {
    background-color: rgb(210, 210, 255);
    border: 3px solid blue;
  }

  &.success {
    background-color: rgb(210, 255, 210);
    border: 3px solid green;
  }

  &.warning {
    background-color: rgb(255, 236, 200);
    border: 3px solid orange;
  }
}

.alertTitle {
  &.error {
    color: red !important;
  }

  &.info {
    color: blue !important;
  }

  &.success {
    color: green !important;
  }

  &.warning {
    color: orange !important;
  }
}

.alertTitleContainer {
  align-items: center;
  display: flex;
  padding-bottom: 3px;
}

.alertIcon {
  align-items: center;
  display: flex;
  font-size: 22px;
  padding-right: 8px;
}

.alertCloseButton {
  cursor: pointer;
  margin-left: 20px;
}

.exitActive {
  opacity: 0;
  transition: all 400ms ease-in-out;
}

.closeButton {
  background-color: transparent !important;
  color: globals.$secondary-text-color !important;
}

.errorButton {
  &:hover,
  &:focus {
    background-color: rgb(255, 180, 180) !important;
  }

  &:active {
    background-color: rgb(255, 150, 150) !important;
  }
}

.infoButton {
  &:hover,
  &:focus {
    background-color: rgb(180, 180, 255) !important;
  }

  &:active {
    background-color: rgb(150, 150, 255) !important;
  }
}

.successButton {
  &:hover,
  &:focus {
    background-color: rgb(180, 255, 180) !important;
  }

  &:active {
    background-color: rgb(150, 255, 150) !important;
  }
}

.warningButton {
  &:hover,
  &:focus {
    background-color: rgb(255, 225, 170) !important;
  }

  &:active {
    background-color: rgb(255, 215, 140) !important;
  }
}
