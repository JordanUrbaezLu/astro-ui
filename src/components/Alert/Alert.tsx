import * as React from "react";
import styles from "./Alert.module.scss";
import classNames from "classnames";
import { CSSTransition } from "react-transition-group";
import { AiOutlineInfoCircle } from "react-icons/ai";
import { BiAlarmExclamation } from "react-icons/bi";
import { FaRegCheckCircle } from "react-icons/fa";
import { FiAlertTriangle, FiX } from "react-icons/fi";
import { IconButton } from "../IconButton";
import { Typography } from "../Typography/Typography";

export type AlertVariant = "error" | "info" | "success" | "warning";

export interface AlertProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The content for the Alert.
   */
  children: string;
  /**
   * If the Alert can close.
   *
   * @default false
   */
  closeable?: boolean;
  /**
   * The variant for the Alert.
   *
   * @default "error"
   */
  variant?: AlertVariant;
}

/**
 * @public
 *
 * @description
 *
 * The Alert component is a notification towards an action a
 * user has recently taken.
 */
export const Alert: React.FC<AlertProps> = (props) => {
  const {
    className,
    children,
    closeable = false,
    variant = "error",
    ...rest
  } = props;

  const alertIcon = (
    <>
      {variant === "error" && <BiAlarmExclamation color="red" />}
      {variant === "info" && <AiOutlineInfoCircle color="blue" />}
      {variant === "success" && <FaRegCheckCircle color="green" />}
      {variant === "warning" && <FiAlertTriangle color="orange" />}
    </>
  );

  const [isOpen, setIsOpen] = React.useState<boolean>(true);

  return (
    <CSSTransition
      classNames={{ exitActive: styles.exitActive }}
      in={isOpen}
      timeout={400}
      mountOnEnter
      unmountOnExit
    >
      <div
        className={classNames(
          className,
          styles.alertContainer,
          variant === "error" && styles.error,
          variant === "info" && styles.info,
          variant === "success" && styles.success,
          variant === "warning" && styles.warning
        )}
        {...rest}
      >
        <div>
          <div className={styles.alertTitleContainer}>
            <div className={styles.alertIcon}>{alertIcon}</div>
            <Typography
              className={classNames(
                styles.alertTitle,
                variant === "error" && styles.error,
                variant === "info" && styles.info,
                variant === "success" && styles.success,
                variant === "warning" && styles.warning
              )}
              type="h2"
              weight="bold"
            >
              {variant?.toUpperCase()}
            </Typography>
          </div>
          <Typography type="h3" variant="secondary">
            {children}
          </Typography>
        </div>
        {closeable && (
          <IconButton
            className={classNames(
              styles.closeButton,
              variant === "error" && styles.errorButton,
              variant === "info" && styles.infoButton,
              variant === "success" && styles.successButton,
              variant === "warning" && styles.warningButton
            )}
            content={<FiX />}
            onClick={() => {
              setIsOpen(false);
            }}
          />
        )}
      </div>
    </CSSTransition>
  );
};

Alert.displayName = "Alert";
