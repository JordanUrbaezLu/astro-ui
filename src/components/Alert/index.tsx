/* istanbul ignore file */
export * from "./Alert";
