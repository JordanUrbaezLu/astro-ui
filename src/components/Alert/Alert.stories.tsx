import { Alert } from "./Alert";
import { FlexColumnContainer } from "../../../.storybook/components/FlexColumnContainer";

export default {
  component: Alert,
  title: "Components/Feedback/Alert",
};

export const Default = () => <Alert>Default Alert!</Alert>;

export const Closeable = () => (
  <Alert closeable variant="info">
    I am a closeable Alert component!
  </Alert>
);

export const Variant = () => {
  return (
    <FlexColumnContainer>
      <Alert variant="error">There was an error!</Alert>
      <Alert variant="info">Here's some information.</Alert>
      <Alert variant="success">That was successful!</Alert>
      <Alert variant="warning">This is a warning!</Alert>
    </FlexColumnContainer>
  );
};

export const VariantAndCloseable = () => {
  return (
    <FlexColumnContainer>
      <Alert closeable variant="error">
        There was an error!
      </Alert>
      <Alert closeable variant="info">
        Here's some information.
      </Alert>
      <Alert closeable variant="success">
        That was successful!
      </Alert>
      <Alert closeable variant="warning">
        This is a warning!
      </Alert>
    </FlexColumnContainer>
  );
};

export const WithLongText = () => (
  <Alert variant="warning">
    I am some very long text for the Alert component!
  </Alert>
);
