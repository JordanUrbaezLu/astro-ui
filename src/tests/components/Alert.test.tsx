import { screen, render, act } from "@testing-library/react";
import { axe } from "jest-axe";
import { Alert, AlertVariant } from "../../components/Alert/Alert";

describe("Alert", () => {
  test("Should render component correctly.", () => {
    const { container } = render(<Alert>This is an alert!</Alert>);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(<Alert>This is an alert!</Alert>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render children correctly.", () => {
    render(<Alert>Alert</Alert>);
    expect(screen.getByText("Alert")).toBeInTheDocument();
  });
  test("Should render close button correctly.", () => {
    render(<Alert closeable>Alert</Alert>);

    const closeButton = screen.getByRole("button");

    expect(closeButton).toBeInTheDocument();

    act(() => closeButton.click());
  });
  test.each<AlertVariant>(["error", "info", "success", "warning"])(
    "Should render variants correctly.",
    (variant) => {
      render(
        <Alert closeable variant={variant}>
          Alert
        </Alert>
      );
      expect(
        document.getElementsByClassName("alertContainer")[0]
      ).toHaveClass(variant);
      expect(screen.getByRole("button")).toBeInTheDocument();
    }
  );
});
