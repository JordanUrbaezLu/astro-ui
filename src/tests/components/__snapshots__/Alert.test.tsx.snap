// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Should render component correctly. 1`] = `
<div
  class="alertContainer error"
>
  <div>
    <div
      class="alertTitleContainer"
    >
      <div
        class="alertIcon"
      >
        <svg
          color="red"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="color: red;"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 22c4.879 0 9-4.121 9-9s-4.121-9-9-9-9 4.121-9 9 4.121 9 9 9zm0-16c3.794 0 7 3.206 7 7s-3.206 7-7 7-7-3.206-7-7 3.206-7 7-7zm5.284-2.293 1.412-1.416 3.01 3-1.413 1.417zM5.282 2.294 6.7 3.706l-2.99 3-1.417-1.413z"
          />
          <path
            d="M11 9h2v5h-2zm0 6h2v2h-2z"
          />
        </svg>
      </div>
      <div
        class="alertTitle error typography typographyh2 primary bold"
      >
        ERROR
      </div>
    </div>
    <div
      class="typography typographyh3 secondary normal"
    >
      This is an alert!
    </div>
  </div>
</div>
`;
